### PR TITLE
Add CollectionsMarshal.SetCount(list, count)

### LINKED ISSUE #55217: Add CollectionsMarshal.SetCount API for List<T>
## Background and Motivation

The existing CollectionsMarshal.AsSpan<T>(List<T>) API allows developers to use List<T> for performant code as a Span, however it restricts modifications of it by having no way to change the Count of the List. Exposing an API in CollectionsMarshal for doing that would solve that and help to remove redundant array allocations and AddRange copies when adding data to a List that comes from an API that takes a Span (AddRange doesn't accept Spans so the buffers can't even be stack allocated).

Implementation details:

- Passing the current count would be an noop
- List version would be changed unless the count would be the same as the current one
- Passing a count bigger than the current capacity, the List would be resized
- Passing a count smaller than the current one would shrink the list and clear the data ONLY if it contains GC references (just like Clear does)
- Passing a count bigger than the current one would increase the count and it would NOT initialize the exposed data.

## Proposed API

```diff
namespace System.Runtime.InteropServices
{
    public static class CollectionsMarshal {
+        public static void SetCount<T>(List<T> list, int count); // alternative could be EnsureCount
    }
}
```

Example implementation:
```C#
public static void SetCount<T>(List<T> list, int count)
{
    if (count < 0)
        throw new ArgumentOutOfRangeException(); // would probably be a throw helper

    if (count == list._size)
        return;

    list._version++; // would need to be made internal

    if (count < list._size)
    {
        if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
            Array.Clear(list._items, count, list._size - count);
        list._size = count;
        return;
    }

    if (count > list.Capacity)
        list.Capacity = count;

    list._size = count;
}
```

## Usage Examples

The API could be used for creating an extension method in place of the missing ReadOnlySpan AddRange:
```C#
public static void AddRange<T>(this List<T> list, ReadOnlySpan<T> span)
{
    int oldCount = list.Count;
    CollectionMarshal.SetCount(list, oldCount + span.Length);
    span.CopyTo(CollectionMarshal.AsSpan(list).Slice(oldCount))
}
```

Random.GetBytes is an example API that fills a Span with data here.
``` C#
List<byte> list = new(16);
CollectionsMarshal.SetCount(list, 16);
Random.Shared.GetBytes(CollectionsMarshal.AsSpan(list));
// we have a count 16 list filled with random data
```

This would be as valid, just a bit less performant:
``` C#
List<byte> list = new();
CollectionsMarshal.SetCount(list, 16);
Random.Shared.GetBytes(CollectionsMarshal.AsSpan(list));
// we have a count 16 list filled with random data
```

With this the end count would also be 16
``` C#
List<byte> list = new() {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20 };
CollectionsMarshal.SetCount(list, 16);
Random.Shared.GetBytes(CollectionsMarshal.AsSpan(list));
// we have a count 16 list filled with random data
```

With this the end count would also be 16 and we'd have the original content from before Clear after SetLength (not guaranteed, implementation detail)
``` C#
List<byte> list = new() {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20 };
list.Clear();
CollectionsMarshal.SetCount(list, 16);
// accessing the list here would give { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 } since Clear wouldn't zero the data
Random.Shared.GetBytes(CollectionsMarshal.AsSpan(list));
// we have a count 16 list filled with random data
```

## Alternative Designs

Currently the developer is required to do this, which requires an array allocation and a copy coming from AddRange:
```C#
List<byte> list = new(16);
list.AddRange(new byte[16]);
```

The first example usage would look like this:
```C#
List<byte> list = new(16);
byte[] array = new byte[16];
RandomNumberGenerator.Fill(array);
list.AddRange(array);
```

## Risks

The API could expose non cleared memory for value types with no GC references (when the List was cleared or if Lists would use GetUninitalizedArray when the initial array was allocated unitialized).

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/55217.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs [hunk 1]
```diff
@@ -39,5 +40,55 @@ public static Span<T> AsSpan<T>(List<T>? list)
...
+                return;
+            }
+
+            list._version++;
+
+            if (count < list._size)
+            {
+                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                {
+                    Array.Clear(list._items, count, list._size - count);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:79]: Why is this needed?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs [hunk 2]
```diff
@@ -39,5 +40,55 @@ public static Span<T> AsSpan<T>(List<T>? list)
...
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="count"/> is negative/>.
+        /// </exception>
+        public static void SetCount<T>(List<T> list, int count)
+        {
+            if (list is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
+            }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:61]: We don't null check dictionary above. Not sure we need to do it here (we do it in AsSpan, but that's mirroring string/array behavior). I don't have a strong opinion, though.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:79]: Oh, size not Capacity, nevermind.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs [hunk 3]
```diff
@@ -39,5 +40,55 @@ public static Span<T> AsSpan<T>(List<T>? list)
...
+
+            list._version++;
+
+            if (count < list._size)
+            {
+                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                {
+                    Array.Clear(list._items, count, list._size - count);
+                    list._size = count;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:80]: Shouldn't this be outside the block?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs [hunk 4]
```diff
@@ -39,5 +40,55 @@ public static Span<T> AsSpan<T>(List<T>? list)
...
+            }
+
+            if (count == list._size)
+            {
+                return;
+            }
+
+            list._version++;
+
+            if (count < list._size)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Is there a use case for setting a value smaller than the current count? That's already possible with RemoveRange.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Would it? Trying to set Capacity to less than Count throws. How is this different? And what's the use case where you'd call this to shrink rather than grow? I'm not against it if it's useful, I'm trying to understand the use.

Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: It'd be useful in cases where you don't know whether the count will be smaller or bigger than the list size.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: What's an example of that?

Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Updating data in an existing list with an API that can return less data.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: I don't follow. Isn't that Clear then AddRange? Can you share a concrete example with code so I can understand a real case where this would be valuable?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Ultimately we have a check here that's either: `if (count < list._size) { throw }` or we have `if (count < list._size) { ClearIfReferenceOrContainsReferences(); list._size = count }`.

Not throwing is less confusing and makes the API more flexible. It's already unsafe and users already need to know what they're doing here.

Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Clear would be less efficient with reference types since it'd zero out all of them, not just the difference. And you can't use AddRange with APIs that take a span without a temporary buffer. That being said, I can't think of a concrete example for now. Will think about it tomorrow.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: It's also worth noting that `Capacity` also supports growing or shrinking, it only doesn't support shrinking below `Count`.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > it only doesn't support shrinking below Count

Right, which is the same situation being discussed here.

> Not throwing is less confusing and makes the API more flexible.

More flexible, sure, less confusing, I'm not convinced. That flexibility has a cost: Capacity doesn't let you drop below the Count because it's lossy. If there aren't known use cases for letting SetCount drop, then I'd rather let the caller know they've made a mistake than silently throw away data.

This API is on CollectionsMarshal, so if there are valid use cases for lowering the count, fine, let's do it... I'd like to understand what those are, though.

> Clear would be less efficient with reference types since it'd zero out all of them, not just the difference

I thought you were suggesting clearing out all of them. If you only want some, then RemoveRange from the end. This is why real scenarios need to be shared, to make it concrete what we're actually enabling, and why.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > Right, which is the same situation being discussed here.

Right, the difference being that setting `Capacity` and having it implicitly modify `Count` seems bad. Where-as the user is directly specifying "set the count" here, so allowing shrinkage can make some sense.

> This API is on CollectionsMarshal, so if there are valid use cases for lowering the count, fine, let's do it... I'd like to understand what those are, though.

The general use-case is the same as for growing it. You want to write custom and efficient algorithms that directly modify the List.

I do think basically every scenario comes down to something that can be achieved with `RemoveRange`. I think the question is really whether we need/want to force users to add additional checks/handling to cover removing "excess data" and whether those will end up forcing algorithms to be more expensive.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > The general use-case is the same as for growing it. You want to write custom and efficient algorithms that directly modify the List.

Can you share a specific real example to help me understand?  All the use cases I've seen have been about growing (e.g. I want to get a span for an appending region I manually fill in, I want to increase the size of the list by X so I can manually poke at the expanded range, etc.).  And in cases where you want to use the list as some sort of efficient pooled growable array, I imagine you'd also possibly not want the clearing overhead when reference types are involved.

Again, I'm not saying we shouldn't allow SetCount to shrink, but there are tradeoffs, and so I want to understand what we're actually getting for it.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: A simple example would actually involve growing the list.

Today, `AsSpan` returns `0, Count` and so if `Capacity > Count`, it won't cover the entire backing array.

You could of course make this more efficient by avoiding bounds checks and the like, but this is a relatively simple sample showing adding an unknown number of items to a list. It's basically a more efficient version of `AddRange` for when you don't need or want to track `version` changes or repeatedly get the fields to account for potential multithreaded access:
```csharp
int currentCount = list.Count;
int currentCapacity = list.Capacity;

// get the entire backing array
CollectionsMarshal.SetCount(list, currentCapacity);
var span = CollectionsMarshal.AsSpan(list);

foreach (var item in items)
{
    if (currentCount < currentCapacity)
    {
        span[currentCount] = item;
    }
    else
    {
        // use the default growth logic
        list.Add(item);

        // get the new backing array
        currentCapacity = list.Capacity;
        CollectionsMarshal.SetCount(list, currentCapacity);

        span = CollectionsMarshal.AsSpan(list);
    }

    currentCount++;
}

// shrink down to the actual count of items
CollectionsMarshal.SetCount(list, currentCount);
```

Part of being able to write custom and efficient algorithms really involves getting a span of the full backing array, making your changes, and then updating the minimal associated metadata (most typically `_count`).

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > A simple example would actually involve growing the list.

Thanks.  But in that example, you may incur significant unnecessary clearing if T is or contains reference types.  I'd think if you cared about efficiency, and if SetCount does clearing, you'd want to avoid this.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Right, I'd personally rather `SetCount` was a very simple API that does the minimum amount of work.

In one of my personal projects I have a `ValueList<T>` implementation, which basically mirrors `List<T>` but is a struct, doesn't track a version, and exposes some helper APIs like `SetCountUnsafe` and `AsSpanUnsafe` directly: https://github.com/terrafx/terrafx/blob/main/sources/Core/Collections/ValueList%601.cs#L327

Such a struct is really meant to be an impl detail and not directly exposed and massively simplifies the ability to create custom collection types (avoids cases where you have `class MyCollection { private List<T> _field; }` and therefore have a triple indirection just to access any item). 

It gets used as some of the backing data stores for my GPU memory manager and allows fast and efficient enumeration and modifications, where appropriate.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:61]: I'm okay with following precedent here & removing the check.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Since we have an existing solution for reducing the count with `RemoveRange`, we can start with throwing now and adjust the behavior if needed based on more user feedback.

Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: Wouldn't such change require the API to be rereviewed?

Comment (DaZombieKiller) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: I agree that `SetCount` throwing would be unexpected. `CollectionsMarshal` is home to several lower-level APIs, so the expectation is that the caller understands what they're doing. Why call it `SetCount` if it only works when you are growing the list?

You can always call `Clear` before `SetCount` to "guarantee" success, but that reads strangely in my opinion:
```cs
if (count < list.Count)
    list.Clear(); // essentially list.SetCount(0)

list.SetCount(count);
```

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > Wouldn't such change require the API to be rereviewed?

Was this behavior discussed at the API review?

> so the expectation is that the caller understands what they're doing

In which case I'm not sure why we're clearing.  If this is about an advanced developer using this unsafe API to do something that requires them to know what they're doing, why are we forcing additional costs upon the operation when we have no idea what the intent is?

Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > Was this behavior discussed at the API review?

Yes, the reviewers agreed on it behaving like RemoveRange.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: I just watched. Alternatives weren't discussed, that was just explaining what the sample code was doing, or how it would behave if this were exposed as the public Count's setter.

Bottom line, I don't see any meaningful scenario for this to be "unsafe" , support shrinking, _and_ spend cycles clearing the shrunken region. But I'll also drop it, because _the_ use case for this method is to increase the count, so, whatever. I think we're making a mistake, but one that probably won't matter.

Comment (DaZombieKiller) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > In which case I'm not sure why we're clearing. If this is about an advanced developer using this unsafe API to do something that requires them to know what they're doing, why are we forcing additional costs upon the operation when we have no idea what the intent is?

I agree. While I don't expect it to be even remotely close to the "common" use case for the API, I _can_ think of scenarios where not having the clearing behavior for reference types could be potentially beneficial.

For example, using a `List<T>` as a temporary buffer for several operations that may resize it, and then at the end you can `SetCount(Capacity)` followed by `Clear` to ensure objects aren't held onto afterwards, rather than potentially having multiple instances of zeroing inside the actual operation.

But the counterpoint is that this makes the API prone to memory leaks `T` is a reference type... I'd argue that's grounds to mark the API with `[UnsafeCallersOnly]` if that existed, but at present I think this is a question of "how unsafe is *too* unsafe for `CollectionsMarshal`".

I personally think that if it's appropriately documented, it should be fine to not have the clearing behavior (and as mentioned, it's expected that callers would understand how the API behaves before using it).

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:75]: > and then at the end you can SetCount(Capacity) followed by Clear to ensure objects aren't held onto afterwards

Sure. But that's just RemoveRange.

> I'd argue that's grounds to mark the API with [UnsafeCallersOnly] if that existed

I expect pretty much everything on `Unsafe`, `Marshal`, `MemoryMarshal`, and `CollectionsMarshal` would be UnsafeCallersOnly if it ever exists.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:70]: In the name of making forward progress here, let's finish this PR with the behavior of supporting shrinking and clearing, and then we can have a follow-up conversation about whether to change the behavior prior to shipping.  Along with this, it'd be good to structure the method in a way that optimizes for the expected case of calling this to increase the count.

Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:70]: This API should be optimized for the cases that do not require buffer resizing, with increasing being the main target. Operations that require a resize could potentially be moved into a separate non inlineable local method.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:70]: Right.

**Comment by MichalPetryka** [comment]
@layomia Is there any progress with this?

**Comment by MichalPetryka** [comment]
@eiriktsarpalis Since @layomia seems to be inactive, would it be fine for me to open a new PR in place of this one?
